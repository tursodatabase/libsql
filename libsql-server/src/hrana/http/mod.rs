use anyhow::{Context, Result};
use bytes::Bytes;
use futures::stream::Stream;
use parking_lot::Mutex;
use serde::{de::DeserializeOwned, Serialize};
use std::pin::Pin;
use std::sync::Arc;
use std::task;

use super::{batch, cursor, Encoding, ProtocolError, Version};
use crate::auth::Authenticated;
use crate::connection::{Connection, MakeConnection};
use crate::hrana::http::stream::StreamError;

mod proto;
mod protobuf;
mod request;
pub(crate) mod stream;

pub struct Server<C> {
    self_url: Option<String>,
    baton_key: [u8; 32],
    stream_state: Mutex<stream::ServerStreamState<C>>,
}

#[derive(Debug, Copy, Clone)]
pub enum Endpoint {
    Pipeline,
    Cursor,
}

impl<C: Connection> Server<C> {
    pub fn new(self_url: Option<String>) -> Self {
        Self {
            self_url,
            baton_key: rand::random(),
            stream_state: Mutex::new(stream::ServerStreamState::new()),
        }
    }

    pub async fn run_expire(&self) {
        stream::run_expire(self).await
    }

    pub async fn handle_request(
        &self,
        connection_maker: Arc<dyn MakeConnection<Connection = C>>,
        auth: Authenticated,
        req: hyper::Request<hyper::Body>,
        endpoint: Endpoint,
        version: Version,
        encoding: Encoding,
    ) -> Result<hyper::Response<hyper::Body>> {
        handle_request(
            self,
            connection_maker,
            auth,
            req,
            endpoint,
            version,
            encoding,
        )
        .await
        .map_err(|e| {
            tracing::error!("hrana server: {}", e);
            e
        })
        .or_else(|err| {
            err.downcast::<StreamError>()
                .map(|err| stream_error_response(err, encoding))
        })
        .or_else(|err| err.downcast::<ProtocolError>().map(protocol_error_response))
    }

    pub(crate) fn stream_state(&self) -> &Mutex<stream::ServerStreamState<C>> {
        &self.stream_state
    }
}

pub(crate) async fn handle_index() -> hyper::Response<hyper::Body> {
    text_response(
        hyper::StatusCode::OK,
        "Hello, this is HTTP API v2 (Hrana over HTTP)".into(),
    )
}

async fn handle_request<C: Connection>(
    server: &Server<C>,
    connection_maker: Arc<dyn MakeConnection<Connection = C>>,
    auth: Authenticated,
    req: hyper::Request<hyper::Body>,
    endpoint: Endpoint,
    version: Version,
    encoding: Encoding,
) -> Result<hyper::Response<hyper::Body>> {
    match endpoint {
        Endpoint::Pipeline => {
            handle_pipeline(server, connection_maker, auth, req, version, encoding).await
        }
        Endpoint::Cursor => {
            handle_cursor(server, connection_maker, auth, req, version, encoding).await
        }
    }
}

async fn handle_pipeline<C: Connection>(
    server: &Server<C>,
    connection_maker: Arc<dyn MakeConnection<Connection = C>>,
    auth: Authenticated,
    req: hyper::Request<hyper::Body>,
    version: Version,
    encoding: Encoding,
) -> Result<hyper::Response<hyper::Body>> {
    let req_body: proto::PipelineReqBody = read_decode_request(req, encoding).await?;
    let mut stream_guard =
        stream::acquire(server, connection_maker, req_body.baton.as_deref()).await?;

    let mut results = Vec::with_capacity(req_body.requests.len());
    for request in req_body.requests.into_iter() {
        tracing::debug!("pipeline:{{ {:?}, {:?} }}", version, request);
        let result = request::handle(&mut stream_guard, auth.clone(), request, version).await?;
        results.push(result);
    }

    let resp_body = proto::PipelineRespBody {
        baton: stream_guard.release(),
        base_url: server.self_url.clone(),
        results,
    };
    Ok(encode_response(hyper::StatusCode::OK, &resp_body, encoding))
}

async fn handle_cursor<C: Connection>(
    server: &Server<C>,
    connection_maker: Arc<dyn MakeConnection<Connection = C>>,
    auth: Authenticated,
    req: hyper::Request<hyper::Body>,
    version: Version,
    encoding: Encoding,
) -> Result<hyper::Response<hyper::Body>> {
    let req_body: proto::CursorReqBody = read_decode_request(req, encoding).await?;
    let stream_guard = stream::acquire(server, connection_maker, req_body.baton.as_deref()).await?;

    let mut join_set = tokio::task::JoinSet::new();
    let mut cursor_hnd = cursor::CursorHandle::spawn(&mut join_set);
    let db = stream_guard.get_db_owned()?;
    let sqls = stream_guard.sqls();
    let pgm = batch::proto_batch_to_program(&req_body.batch, sqls, version)?;
    cursor_hnd.open(db, auth, pgm, req_body.batch.replication_index);

    let resp_body = proto::CursorRespBody {
        baton: stream_guard.release(),
        base_url: server.self_url.clone(),
    };
    let body = hyper::Body::wrap_stream(CursorStream {
        resp_body: Some(resp_body),
        join_set,
        cursor_hnd,
        encoding,
    });
    let content_type = match encoding {
        Encoding::Json => "text/plain",
        Encoding::Protobuf => "application/octet-stream",
    };

    Ok(hyper::Response::builder()
        .status(hyper::StatusCode::OK)
        .header(hyper::http::header::CONTENT_TYPE, content_type)
        .body(body)
        .unwrap())
}

struct CursorStream<D> {
    resp_body: Option<proto::CursorRespBody>,
    join_set: tokio::task::JoinSet<()>,
    cursor_hnd: cursor::CursorHandle<D>,
    encoding: Encoding,
}

impl<D> Stream for CursorStream<D> {
    type Item = Result<Bytes>;

    fn poll_next(
        self: Pin<&mut Self>,
        cx: &mut task::Context,
    ) -> task::Poll<Option<Result<Bytes>>> {
        let this = self.get_mut();

        if let Some(resp_body) = this.resp_body.take() {
            let chunk = encode_stream_item(&resp_body, this.encoding);
            return task::Poll::Ready(Some(Ok(chunk)));
        }

        match this.join_set.poll_join_next(cx) {
            task::Poll::Pending => {}
            task::Poll::Ready(Some(Ok(()))) => {}
            task::Poll::Ready(Some(Err(err))) => panic!("Cursor task crashed: {}", err),
            task::Poll::Ready(None) => {}
        };

        match this.cursor_hnd.poll_fetch(cx) {
            task::Poll::Pending => task::Poll::Pending,
            task::Poll::Ready(None) => task::Poll::Ready(None),
            task::Poll::Ready(Some(Ok(entry))) => {
                let chunk = encode_stream_item(&entry.entry, this.encoding);
                task::Poll::Ready(Some(Ok(chunk)))
            }
            task::Poll::Ready(Some(Err(err))) => task::Poll::Ready(Some(Err(err))),
        }
    }
}

fn encode_stream_item<T: Serialize + prost::Message>(item: &T, encoding: Encoding) -> Bytes {
    let mut data: Vec<u8>;
    match encoding {
        Encoding::Json => {
            data = serde_json::to_vec(item).unwrap();
            data.push(b'\n');
        }
        Encoding::Protobuf => {
            data = <T as prost::Message>::encode_length_delimited_to_vec(item);
        }
    }
    Bytes::from(data)
}

async fn read_decode_request<T: DeserializeOwned + prost::Message + Default>(
    req: hyper::Request<hyper::Body>,
    encoding: Encoding,
) -> Result<T> {
    let req_body = hyper::body::to_bytes(req.into_body())
        .await
        .context("Could not read request body")?;
    match encoding {
        Encoding::Json => serde_json::from_slice(&req_body)
            .map_err(|err| ProtocolError::JsonDeserialize { source: err })
            .context("Could not deserialize JSON request body"),
        Encoding::Protobuf => <T as prost::Message>::decode(req_body)
            .map_err(|err| ProtocolError::ProtobufDecode { source: err })
            .context("Could not decode Protobuf request body"),
    }
}

fn protocol_error_response(err: ProtocolError) -> hyper::Response<hyper::Body> {
    text_response(hyper::StatusCode::BAD_REQUEST, err.to_string())
}

fn stream_error_response(err: StreamError, encoding: Encoding) -> hyper::Response<hyper::Body> {
    let status = match err {
        StreamError::StreamExpired => hyper::StatusCode::BAD_REQUEST,
    };
    encode_response(
        status,
        &proto::Error {
            message: err.to_string(),
            code: err.code().into(),
        },
        encoding,
    )
}

fn encode_response<T: Serialize + prost::Message>(
    status: hyper::StatusCode,
    resp_body: &T,
    encoding: Encoding,
) -> hyper::Response<hyper::Body> {
    let (resp_body, content_type) = match encoding {
        Encoding::Json => (serde_json::to_vec(resp_body).unwrap(), "application/json"),
        Encoding::Protobuf => (
            <T as prost::Message>::encode_to_vec(resp_body),
            "application/x-protobuf",
        ),
    };
    hyper::Response::builder()
        .status(status)
        .header(hyper::http::header::CONTENT_TYPE, content_type)
        .body(hyper::Body::from(resp_body))
        .unwrap()
}

fn text_response(status: hyper::StatusCode, resp_body: String) -> hyper::Response<hyper::Body> {
    hyper::Response::builder()
        .status(status)
        .header(hyper::http::header::CONTENT_TYPE, "text/plain")
        .body(hyper::Body::from(resp_body))
        .unwrap()
}
