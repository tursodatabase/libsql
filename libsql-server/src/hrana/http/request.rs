use anyhow::{anyhow, bail, Result};
use bytesize::ByteSize;

use super::super::{batch, stmt, ProtocolError, Version};
use super::stream;
use crate::connection::{Connection, RequestContext};
use libsql_hrana::proto;

const MAX_SQL_COUNT: usize = 50;
const MAX_STORED_SQL_SIZE: ByteSize = ByteSize::kb(5);

/// An error from executing a [`proto::StreamRequest`]
#[derive(thiserror::Error, Debug)]
enum StreamResponseError {
    #[error("The server already stores {count} SQL texts, it cannot store more")]
    SqlTooMany { count: usize },
    #[error("The statement is too large to be stored: {size}/{MAX_STORED_SQL_SIZE}")]
    SqlTooLarge { size: ByteSize },
    #[error(transparent)]
    Stmt(stmt::StmtError),
    #[error(transparent)]
    Batch(batch::BatchError),
}

pub async fn handle(
    stream_guard: &mut stream::Guard<'_>,
    ctx: RequestContext,
    request: proto::StreamRequest,
    version: Version,
) -> Result<proto::StreamResult> {
    let result = match try_handle(stream_guard, ctx, request, version).await {
        Ok(response) => proto::StreamResult::Ok { response },
        Err(err) => {
            let resp_err = err.downcast::<StreamResponseError>()?;
            let error = proto::Error {
                message: resp_err.to_string(),
                code: resp_err.code().into(),
            };
            proto::StreamResult::Error { error }
        }
    };
    Ok(result)
}

async fn try_handle(
    stream_guard: &mut stream::Guard<'_>,
    ctx: RequestContext,
    request: proto::StreamRequest,
    version: Version,
) -> Result<proto::StreamResponse> {
    macro_rules! ensure_version {
        ($min_version:expr, $what:expr) => {
            if version < $min_version {
                bail!(ProtocolError::NotSupported {
                    what: $what,
                    min_version: $min_version,
                })
            }
        };
    }

    Ok(match request {
        proto::StreamRequest::None => bail!(ProtocolError::NoneStreamRequest),
        proto::StreamRequest::Close(_req) => {
            stream_guard.close_db();
            proto::StreamResponse::Close(proto::CloseStreamResp {})
        }
        proto::StreamRequest::Execute(req) => {
            let db = stream_guard.get_db()?;
            let sqls = stream_guard.sqls();
            let query =
                stmt::proto_stmt_to_query(&req.stmt, sqls, version).map_err(catch_stmt_error)?;
            let result = stmt::execute_stmt(db, ctx, query, req.stmt.replication_index)
                .await
                .map_err(catch_stmt_error)?;
            proto::StreamResponse::Execute(proto::ExecuteStreamResp { result })
        }
        proto::StreamRequest::Batch(req) => {
            let db = stream_guard.get_db()?;
            let sqls = stream_guard.sqls();
            let pgm = batch::proto_batch_to_program(&req.batch, sqls, version)
                .map_err(catch_stmt_error)
                .map_err(catch_batch_error)?;
            let result = batch::execute_batch(db, ctx, pgm, req.batch.replication_index)
                .await
                .map_err(catch_batch_error)?;
            proto::StreamResponse::Batch(proto::BatchStreamResp { result })
        }
        proto::StreamRequest::Sequence(req) => {
            let db = stream_guard.get_db()?;
            let sqls = stream_guard.sqls();
            let sql = stmt::proto_sql_to_sql(req.sql.as_deref(), req.sql_id, sqls, version)?;
            let pgm = batch::proto_sequence_to_program(sql).map_err(catch_stmt_error)?;
            batch::execute_sequence(db, ctx, pgm, req.replication_index)
                .await
                .map_err(catch_stmt_error)
                .map_err(catch_batch_error)?;
            proto::StreamResponse::Sequence(proto::SequenceStreamResp {})
        }
        proto::StreamRequest::Describe(req) => {
            let db = stream_guard.get_db()?;
            let sqls = stream_guard.sqls();
            let sql = stmt::proto_sql_to_sql(req.sql.as_deref(), req.sql_id, sqls, version)?;
            let result = stmt::describe_stmt(db, ctx, sql.into(), req.replication_index)
                .await
                .map_err(catch_stmt_error)?;
            proto::StreamResponse::Describe(proto::DescribeStreamResp { result })
        }
        proto::StreamRequest::StoreSql(req) => {
            let sqls = stream_guard.sqls_mut();
            let sql_id = req.sql_id;
            if sqls.contains_key(&sql_id) {
                bail!(ProtocolError::SqlExists { sql_id })
            } else if sqls.len() >= MAX_SQL_COUNT {
                bail!(StreamResponseError::SqlTooMany { count: sqls.len() })
            } else if req.sql.len() > MAX_STORED_SQL_SIZE.as_u64() as usize {
                bail!(StreamResponseError::SqlTooLarge {
                    size: ByteSize::b(req.sql.len() as _)
                })
            }
            sqls.insert(sql_id, req.sql);
            proto::StreamResponse::StoreSql(proto::StoreSqlStreamResp {})
        }
        proto::StreamRequest::CloseSql(req) => {
            let sqls = stream_guard.sqls_mut();
            sqls.remove(&req.sql_id);
            proto::StreamResponse::CloseSql(proto::CloseSqlStreamResp {})
        }
        proto::StreamRequest::GetAutocommit(_req) => {
            ensure_version!(Version::Hrana3, "The `get_autocommit` request");
            let db = stream_guard.get_db()?;
            let is_autocommit = db.is_autocommit().await?;
            proto::StreamResponse::GetAutocommit(proto::GetAutocommitStreamResp { is_autocommit })
        }
    })
}

fn catch_stmt_error(err: anyhow::Error) -> anyhow::Error {
    match err.downcast::<stmt::StmtError>() {
        Ok(stmt_err) => anyhow!(StreamResponseError::Stmt(stmt_err)),
        Err(err) => match err.downcast::<crate::Error>() {
            Ok(crate::Error::Migration(crate::schema::Error::MigrationError(_step, message))) => {
                anyhow!(StreamResponseError::Stmt(stmt::StmtError::SqliteError {
                    source: rusqlite::ffi::Error {
                        code: rusqlite::ffi::ErrorCode::Unknown,
                        extended_code: 4242
                    },
                    message
                }))
            }
            Ok(err) => anyhow!(err),
            Err(err) => err,
        },
    }
}

fn catch_batch_error(err: anyhow::Error) -> anyhow::Error {
    match err.downcast::<batch::BatchError>() {
        Ok(batch_err) => anyhow!(StreamResponseError::Batch(batch_err)),
        Err(err) => err,
    }
}

impl StreamResponseError {
    pub fn code(&self) -> &'static str {
        match self {
            Self::SqlTooMany { .. } => "SQL_STORE_TOO_MANY",
            Self::Stmt(err) => err.code(),
            Self::Batch(err) => err.code(),
            StreamResponseError::SqlTooLarge { .. } => "SQL_STORE_TOO_LARGE",
        }
    }
}
