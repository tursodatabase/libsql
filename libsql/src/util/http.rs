use super::box_clone_service::BoxCloneService;
use tokio::io::{AsyncRead, AsyncWrite};

pub trait Socket:
    hyper::client::connect::Connection + AsyncRead + AsyncWrite + Send + Unpin + 'static + Sync
{
}

impl<T> Socket for T where
    T: hyper::client::connect::Connection + AsyncRead + AsyncWrite + Send + Unpin + 'static + Sync
{
}

impl hyper::client::connect::Connection for Box<dyn Socket> {
    fn connected(&self) -> hyper::client::connect::Connected {
        self.as_ref().connected()
    }
}

pub type ConnectorService =
    BoxCloneService<http::Uri, Box<dyn Socket>, Box<dyn std::error::Error + Sync + Send + 'static>>;

#[cfg(feature = "replication")]
pub type HttpRequestCallback = std::sync::Arc<dyn Fn(&mut http::Request<()>) + Send + Sync>;
