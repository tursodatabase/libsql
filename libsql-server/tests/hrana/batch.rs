use std::time::Duration;

use insta::assert_json_snapshot;
use libsql::{params, Database};
use libsql_server::hrana_proto::{Batch, BatchStep, Stmt};

use crate::common::http::Client;
use crate::common::net::TurmoilConnector;

#[test]
fn sample_request() {
    let mut sim = turmoil::Builder::new()
        .simulation_duration(Duration::from_secs(1000))
        .build();
    sim.host("primary", super::make_standalone_server);
    sim.client("client", async {
        let batch = Batch {
            steps: vec![BatchStep {
                condition: None,
                stmt: Stmt {
                    sql: Some("create table test (x)".to_string()),
                    ..Default::default()
                },
            }],
            replication_index: None,
        };
        let client = Client::new();

        let resp = client
            .post(
                "http://primary:8080/v1/batch",
                serde_json::json!({ "batch": batch }),
            )
            .await
            .unwrap();

        let mut json = resp.json_value().await.unwrap();

        for result in json["result"]["step_results"]
            .as_array_mut()
            .unwrap()
            .iter_mut()
        {
            result
                .as_object_mut()
                .unwrap()
                .remove("query_duration_ms")
                .expect("expected query_duration_ms");
        }

        assert_json_snapshot!(json);

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
fn execute_individual_statements() {
    let mut sim = turmoil::Builder::new()
        .simulation_duration(Duration::from_secs(1000))
        .build();
    sim.host("primary", super::make_standalone_server);
    sim.client("client", async {
        let db = Database::open_remote_with_connector("http://primary:8080", "", TurmoilConnector)?;
        let conn = db.connect()?;

        conn.execute("create table t(x text)", ()).await?;
        conn.execute("insert into t(x) values(?)", params!["hello"])
            .await?;
        let mut rows = conn
            .query("select * from t where x = ?", params!["hello"])
            .await?;

        assert_eq!(rows.column_count(), 1);
        assert_eq!(rows.column_name(0), Some("x"));
        assert_eq!(rows.next().await?.unwrap().get::<String>(0)?, "hello");
        assert!(rows.next().await?.is_none());

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
fn execute_batch() {
    let mut sim = turmoil::Builder::new()
        .simulation_duration(Duration::from_secs(1000))
        .build();
    sim.host("primary", super::make_standalone_server);
    sim.client("client", async {
        let db = Database::open_remote_with_connector("http://primary:8080", "", TurmoilConnector)?;
        let conn = db.connect()?;

        conn.execute_batch(
            r#"
        begin;
        create table t(x text);
        insert into t(x) values('hello; world');
        end;"#,
        )
        .await?;
        let mut rows = conn
            .query("select * from t where x = ?", params!["hello; world"])
            .await?;

        assert_eq!(rows.column_count(), 1);
        assert_eq!(rows.column_name(0), Some("x"));
        assert_eq!(
            rows.next().await?.unwrap().get::<String>(0)?,
            "hello; world"
        );
        assert!(rows.next().await?.is_none());

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
fn multistatement_query() {
    let mut sim = turmoil::Builder::new()
        .simulation_duration(Duration::from_secs(1000))
        .build();
    sim.host("primary", super::make_standalone_server);
    sim.client("client", async {
        let db = Database::open_remote_with_connector("http://primary:8080", "", TurmoilConnector)?;
        let conn = db.connect()?;
        let mut rows = conn
            .query("select 1 + ?; select 'abc';", params![1])
            .await?;

        assert_eq!(rows.column_count(), 1);
        assert_eq!(rows.next().await?.unwrap().get::<i32>(0)?, 2);
        assert!(rows.next().await?.is_none());

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
fn affected_rows_and_last_rowid() {
    let mut sim = turmoil::Builder::new()
        .simulation_duration(Duration::from_secs(1000))
        .build();
    sim.host("primary", super::make_standalone_server);
    sim.client("client", async {
        let db = Database::open_remote_with_connector("http://primary:8080", "", TurmoilConnector)?;
        let conn = db.connect()?;

        conn.execute(
            "create table t(id integer primary key autoincrement, x text);",
            (),
        )
        .await?;

        let r = conn.execute("insert into t(x) values('a');", ()).await?;
        assert_eq!(r, 1, "1st row inserted");
        assert_eq!(conn.last_insert_rowid(), 1, "1st row id");

        let r = conn
            .execute("insert into t(x) values('b'),('c');", ())
            .await?;
        assert_eq!(r, 2, "2nd and 3rd rows inserted");
        assert_eq!(conn.last_insert_rowid(), 3, "3rd row id");

        let r = conn.execute("update t set x = 'd';", ()).await?;
        assert_eq!(r, 3, "all three rows updated");
        assert_eq!(conn.last_insert_rowid(), 3, "last row id unchanged");

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
fn stats() {
    let mut sim = turmoil::Builder::new()
        .simulation_duration(Duration::from_secs(1000))
        .build();
    sim.host("primary", super::make_standalone_server);
    sim.client("client", async {
        let req = serde_json::json!({
            "requests": [
                {"type": "execute", "stmt": { "sql": "CREATE TABLE foo (x INT)" }},
                {"type": "execute", "stmt": { "sql": "INSERT INTO foo VALUES (42)"}},
                {"type": "execute", "stmt": { "sql": "SELECT * FROM foo"}}
            ]
        });
        let client = Client::new();

        let resp = client
            .post("http://primary:8080/v2/pipeline", req)
            .await
            .unwrap();

        let mut json = resp.json_value().await.unwrap();
        json.as_object_mut().unwrap().remove("baton");

        for results in json["results"].as_array_mut().unwrap().iter_mut() {
            results["response"]["result"]
                .as_object_mut()
                .unwrap()
                .remove("query_duration_ms")
                .expect("expected query_duration_ms");
        }

        assert_json_snapshot!(json);

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
fn stats_legacy() {
    let mut sim = turmoil::Builder::new()
        .simulation_duration(Duration::from_secs(1000))
        .build();
    sim.host("primary", super::make_standalone_server);
    sim.client("client", async {
        let req = serde_json::json!({
            "statements": [
                "CREATE TABLE foo (x INT)",
                "INSERT INTO foo VALUES (42)",
                "SELECT * FROM foo"
            ]
        });
        let client = Client::new();

        let resp = client.post("http://primary:8080/", req).await.unwrap();

        let mut json = resp.json_value().await.unwrap();

        for result in json.as_array_mut().unwrap() {
            result["results"]
                .as_object_mut()
                .unwrap()
                .remove("query_duration_ms")
                .expect("expected query_duration_ms");
        }

        assert_json_snapshot!(json);

        Ok(())
    });

    sim.run().unwrap();
}
