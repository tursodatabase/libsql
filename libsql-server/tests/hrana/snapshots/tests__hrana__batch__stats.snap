---
source: libsql-server/tests/hrana/batch.rs
expression: json
---
{
  "results": [
    {
      "type": "ok",
      "response": {
        "type": "execute",
        "result": {
          "cols": [],
          "rows": [],
          "affected_row_count": 0,
          "last_insert_rowid": null,
          "replication_index": "1",
          "rows_read": 1,
          "rows_written": 2
        }
      }
    },
    {
      "type": "ok",
      "response": {
        "type": "execute",
        "result": {
          "cols": [],
          "rows": [],
          "affected_row_count": 1,
          "last_insert_rowid": "1",
          "replication_index": "2",
          "rows_read": 0,
          "rows_written": 1
        }
      }
    },
    {
      "type": "ok",
      "response": {
        "type": "execute",
        "result": {
          "cols": [
            {
              "name": "x",
              "decltype": "INT"
            }
          ],
          "rows": [
            [
              {
                "type": "integer",
                "value": "42"
              }
            ]
          ],
          "affected_row_count": 0,
          "last_insert_rowid": null,
          "replication_index": "2",
          "rows_read": 1,
          "rows_written": 0
        }
      }
    }
  ],
  "base_url": null
}
